### DISCUSSION TITLE
Flux

### DISCUSSION THREAD

**Discussion by Venoox** [mysticaltech/terraform-hcloud-kube-hetzner#616] [Q&A] [answered]
Has anyone successfully bootstrapped Flux with this template?

**Comment by mysticaltech** [comment]
Yes, flux is easy to install via helm if I am not mistaken.

**Comment by 89Q12** [comment, marked as answer]
Hi @Venoox, via helm is possible as @mysticaltech wrote or with the provided CLI, but you want to use Kustomize for the optimal workflow.
For a complete automated bootstrap of the cluster including Flux, you may want to follow my setup shown below!
# Ground work:
You need to create a folder named `extra-manifests`, as outlined in the README for Kustomize to be executed after initial K3S installation and you need the Flux CLI, for creating the secret mentioned later.
# Kustomization time:
Create a file named `kustomization.yaml.tpl` in the folder created above, with this content:
```
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
bases:
- https://raw.githubusercontent.com/fluxcd/flux2/main/manifests/install/kustomization.yaml
resources:
- flux-secret.yaml
- /path/to/the/flux/kustomization.yaml
- /path/to/the/flux/git-repo
```
Don't worry about the resources yet other than adding the correct path to them, we will come to that below.
If you already have them you can ignore them.

# Secrets and more files
Now you have the basic structure, you need 3 more important files:
- The secret for Flux that contains 2 keys:
    - identity
    - identity.pub
- The GitRepository definition, so that Flux knows where to look for changes.
- The Flux Kustomization, that points to the directory with all the deployments etc
  Now creating the secret could be done manually, but I will refer to using the Flux CLI.
  Using the CLI you need to execute this command:
```
  flux create secret git flux-system \
--url=ssh://git@github.com/username/repoName \
--export > flux-secret.yaml
```
This will generate a new SSH key and put it in the secret, you DO NOT want to push this to the git repository.
Now you're almost done with the Secret, only thing left is to add the public key from the secret to your choosen Github repository.
`cat flux-secret.yaml` the copy the data under the key `identity.pub`.

# Flux Kustomization and GitRepository
Above when creating the` kustomization.yam.tpl`, you already added paths to the two missing files, now we need to create them if you don't have them already.
To create them we are going to use the CLI again:
```
flux create source git flux-system \
 --url=ssh://git@github.com/username/repoName \
 --branch=master \
 --secret-ref flux-system \
 --export > ../clusters/production/gotk-sync.yaml \
&& \
flux create kustomization flux-system \
 --source=GitRepository/flux-system \
 --path="./apps/production/" \
 --prune=true \
 --interval=10m \
 --validation=client \
 --export >> ../clusters/production/flux-system/gotk-sync.yaml
```
 This assumes a repo structure as shown in example repository of Flux, you can change this obviously.
 
 # Finishing up
 
 You should now have a folder named `extrax_manifests`, which contains:
- flux-secret.yaml
- kustomization.yaml.tpl

And in your the final Flux folder you should have:
- gotk-sync.yaml

This should do the trick of automatically deploying Flux to your newly created cluster after executing `terraform apply`

I hope this helps and solves your issue, if you have questions feel free to ping me! 
Thanks for reading :3

**Reply by mysticaltech** [reply]
@11Tuvork28 Superb explanation, thanks!
